cellxgene_schema_cli/cellxgene_schema/validate.py

```diff
@@ -1132,21 +1132,46 @@ def _is_raw(self, force: bool = False) -> bool:
             # Get potential raw_X
             raw_loc = self._get_raw_x_loc()
             x = self.adata.raw.X if raw_loc == "raw.X" else self.adata.X
+            if x.dtype != np.float32:
+                self._raw_layer_exists = False
+                self.errors.append("Raw matrix values must have type numpy.float32.")
+                return self._raw_layer_exists
 
             matrix_format = get_matrix_format(self.adata, x)
             assert matrix_format != "unknown"
             self._raw_layer_exists = True
+            has_nonzero_in_every_row = True
+            all_nonzero_valid = True
             for matrix_chunk, _, _ in self._chunk_matrix(x):
-                data = matrix_chunk if isinstance(matrix_chunk, np.ndarray) else matrix_chunk.data
-                if (data % 1 > 0).any():
-                    self._raw_layer_exists = False
+
+                if has_nonzero_in_every_row:
+                    if matrix_format in ("csr", "csc", "coo"):  # full set of sparse matrix types
+                        row_indices, _ = matrix_chunk.nonzero()
+                        if len(set(row_indices)) != self.adata.n_obs:
+                            has_nonzero_in_every_row = False
+                    elif matrix_format == "dense" and not all(np.apply_along_axis(np.any, axis=1, arr=matrix_chunk)):
+                        has_nonzero_in_every_row = False
+
+                if all_nonzero_valid:
+                    data = matrix_chunk if isinstance(matrix_chunk, np.ndarray) else matrix_chunk.data
+                    if np.any((data % 1 > 0) | (data < 0)):
+                        all_nonzero_valid = False
+
+                if not has_nonzero_in_every_row and not all_nonzero_valid:
```

Comment:
This is a very logic heavy statement with double negative (not nonzero). It would be easier to understand with positive logic. If has a value in every row and all values are valid. Value is kind of vague and doesn't imply not zero. We could also try reversing the logic in the `has_nonzero_in_every_row = True` to `has_row_of_zeros = False` same goes for `all_nonzero_valid`->`invalid_nonzero_value`

```suggestion
                if has_row_of_zeros or invalid_nonzero_value):
```

Reply:
re-worked, should be clearer now